Implement member type check in interfaces

LemonBoy opened the pull request:
Most of the code is a small refactoring to allow the use of a where_T for signaling the type mismatch, the type checking itself is pretty simple.

LemonBoy committed:
Implement member type check in interfaces

Most of the code is a small refactoring to allow the use of a where_T
for signaling the type mismatch, the type checking itself is pretty
simple.

yegappan commented on src/vim9type.c:
```diff
@@ -744,15 +744,20 @@ type_mismatch_where(type_T *expected, type_T *actual, where_T where)
     {
 	if (where.wt_func_name == NULL)
-	    semsg(_(where.wt_variable
+	    semsg(_(where.wt_kind
```
Is it better to use `where.wt_kind != WT_ARGUMENT` here?
or change the check to `where.wt_kind == WT_ARGUMENT` and reverse the values?

yegappan commented on src/vim9type.c:
```diff
@@ -744,15 +744,20 @@ type_mismatch_where(type_T *expected, type_T *actual, where_T where)
 					 where.wt_index, typename1, typename2);
 	else
-	    semsg(_(where.wt_variable
+	    semsg(_(where.wt_kind
```
Is it better to use `where.wt_kind != WT_ARGUMENT` here?
or change the check to `where.wt_kind == WT_ARGUMENT` and reverse the values?

LemonBoy committed:
Improve where_T error reports

Let the caller explicitly define the kind of location it's referring to
and free the WT_ARGUMENT enum from its catch-all role.

LemonBoy committed:
Implement type checking for interface methods

Follows closely the logic used for type-checking the members.
